File: 101/ruby/tree.rb
# frozen_string_literal: true

# Binary search tree implementation.
class BinarySearchTree
  # Create a new, empty tree.
  def initialize
    @root = nil
  end

  # Add a new value to the tree.
  def add(value)
    @root = add_to_node(@root, value)
  end

  # Traverse the tree in a breadth first manner.
  def breadth_first_traversal(node = @root)
    queue = []
    queue.push(node)

    until queue.empty?
      current = queue.shift
      yield(current.value)
      queue.push(current.left) unless current.left.nil?
      queue.push(current.right) unless current.right.nil?
    end
  end

  # Check if the tree contains a specified value.
  def contains?(value, node = @root)
    return false if node.nil?

    if node.value == value
      true
    elsif value > node.value
      contains?(value, node.right)
    else
      contains?(value, node.left)
    end
  end

  # Traverse the tre in a depth first manner.
  def depth_first_traversal(node = @root, &block)
    return if node.nil?

    depth_first_traversal(node.left, &block)
    yield(node.value)
    depth_first_traversal(node.right, &block)
  end

  # Remove a value from the tree, if it exists.
  # rubocop:disable Metrics/AbcSize
  # rubocop:disable Metrics/MethodLength
  # rubocop:disable Metrics/PerceivedComplexity
  def remove(value, node = @root)
    return node if node.nil?

    if node.value == value
      if node.leaf?
        nil
      elsif node.left.nil?
        node.right
      elsif node.right.nil?
        node.left
      else
        smallest = find_smallest(node.right)
        node.value = smallest
        node.right = remove(smallest, node.right)
        node
      end
    elsif value > node.value
      node.right = remove(value, node.right)
    else
      node.left = remove(value, node.left)
    end
  end
  # rubocop:enable Metrics/AbcSize
  # rubocop:enable Metrics/MethodLength
  # rubocop:enable Metrics/PerceivedComplexity

  private

  # Add a value to a specified node. Recurses the tree to find the correct
  # position in the tree.
  def add_to_node(node, value)
    return TreeNode.new(value) if node.nil?
    return node if value == node.value

    if value < node.value
      node.left = add_to_node(node.left, value)
    else
      node.right = add_to_node(node.right, value)
    end

    node
  end

  # Find the smallest item amongst a branch.
  def find_smallest(node)
    if node.left.nil?
      node.value
    else
      find_smallest(node.right)
    end
  end
end

# Represents a node in a binary tree.
class TreeNode
  attr_accessor :value, :left, :right

  # Create a new tree node with a value and, optionally, left and right
  # child nodes.
  def initialize(value, left = nil, right = nil)
    @value = value
    @left = left
    @right = right
  end

  # Check if a node is a leaf node
  def leaf?
    @left.nil? && right.nil?
  end
end

File: 101/ts/src/tree.ts
class Node {
  value: number;
  left?: Node;
  right?: Node;

  /**
   * Represents a node in the tree.
   *
   * @param {Number} providedValue the value for the node.
   * @param {Node} providedLeft initial value for the left child node of the tree.
   * @param {Node} providedRight initial value for the right child node of the tree.
   * @constructor
   */
  constructor(value: number, left?: Node, right?: Node) {
    this.value = value;
    this.left = left;
    this.right = right;
  }

  isLeaf = () => !this.left && !this.right;
}

export class SortedBinaryTree {
  root?: Node;

  /**
   * Add an element to the binary tree.
   *
   * @param {Number} value value to add to the binary tree.
   */
  add = (value: number): void => {
    this.root = this._add(value, this.root);
  };

  /**
   * Perform a depth-first search for a value. As the tree is sorted there is
   * not much reason for any other type of search.
   */
  contains = (value: number): boolean => this.search(value, this.root);

  /**
   * Traverse the tree in a breadth first manner.
   *
   * @param {Function} callback method to call with each value as it is traversed.
   */
  breadthFirstTraverse = (callback: (value: number) => void): void =>
    this.bf(callback, this.root);

  /**
   * Traverse the tree in a breadth first manner.
   *
   * @param {Function} callback method to call with each value as it is traversed.
   */
  depthFirstTraverse = (callback: (value: number) => void): void =>
    this.df(callback, this.root);

  /**
   * Remove an element from the tree.
   *
   * @param {Number} value the value to remove from the tree.
   */
  remove = (value: number): boolean => !!this.delete(value, this.root);

  /**
   * Recursively find the element where the value should be added.
   *
   * @param {Node} to the current node in the traversal.
   * @param {Number} value value to add to the binary tree.
   */
  private _add = (value: number, to?: Node): Node => {
    if (!to) {
      return new Node(value);
    }

    if (value < to.value) {
      to.left = this._add(value, to.left);
    } else if (value > to.value) {
      to.right = this._add(value, to.right);
    }

    return to;
  };

  /**
   * Recursively traverse a node in a breadth-first manner.
   *
   * @param {Node} node the node to traverse.
   * @param {Function} callback the function to call with each value as
   *                            we traverse the node.
   */
  bf = (callback: (value: number) => void, node?: Node): void => {
    if (node) {
      const queue = new Array<Node>();
      queue.push(node);

      while (queue.length > 0) {
        const current = queue.shift();
        if (current) {
          callback(current.value);

          if (current.left) {
            queue.push(current.left);
          }
          if (current.right) {
            queue.push(current.right);
          }
        }
      }
    }
  };

  /**
   * Traverse the tree recursively and delete the item specified, if found.
   *
   * @param {Node} from the node to start the search from.
   * @param {Number} value the value to remove.
   */
  private delete = (value: number, from?: Node): Node | undefined => {
    if (!from) {
      return undefined;
    }

    if (from.value === value) {
      if (from.isLeaf()) {
        return undefined;
      } else if (!from.right) {
        return from.left;
      } else if (!from.left) {
        return from.right;
      } else {
        const smallestValue = this.findSmallest(from.right);
        from.value = smallestValue;
        from.right = this.delete(smallestValue, from.right);
        return from;
      }
    }

    if (value < from.value) {
      return this.delete(value, from.left);
    } else {
      return this.delete(value, from.right);
    }
  };

  /**
   * Recursively traverse a node in a depth-first manner.
   *
   * @param {Node} node the node to traverse.
   * @param {Function} callback the function to call with each value as
   *                            we traverse the node.
   */
  private df = (callback: (value: number) => void, node?: Node) => {
    if (node) {
      this.df(callback, node.left);
      callback(node.value);
      this.df(callback, node.right);
    }
  };

  /**
   * Perform a depth-first search for a value.
   *
   * @param {Node} node the node to search.
   * @param {Number} value the value to search for.
   */
  private search = (value: number, node?: Node): boolean => {
    if (!node) {
      return false;
    }

    if (node.value === value) {
      return true;
    }

    if (value < node.value) {
      return this.search(value, node.left);
    } else {
      return this.search(value, node.right);
    }
  };

  /**
   * Find the smallest value in a branch.
   *
   * @param {Node} from the node to traverse.
   */
  private findSmallest = (from: Node): number =>
    from.left ? this.findSmallest(from.left) : from.value;
}
